src/main/java/htsjdk/samtools/util/GZIIndex.java

```diff
@@ -0,0 +1,365 @@
+        if (output == null) {
+            throw new IllegalArgumentException("null output path");
+        }
+
+        // the first entry is never written - 0, 0
+        final int numberOfBlocksToWrite = getNumberOfBlocks() - 1;
+
+        final ByteBuffer buffer = allocateBuffer(numberOfBlocksToWrite, true);
+        // put the number of entries
+        buffer.putLong(Integer.toUnsignedLong(numberOfBlocksToWrite));
+
+        // except the first, iterate over all the entries
+        for (final IndexEntry entry : entries.subList(1, entries.size())) {
+            // implementation of entry ensures that the offsets are no negative
+            buffer.putLong(entry.getCompressedOffset());
+            buffer.putLong(entry.getUncompressedOffset());
+        }
+
+        // write into the output
+        Files.write(output, buffer.array());
+    }
+
+    /**
+     * Loads the index from the provided file.
+     *
+     * @param indexPath the path for the index to load.
+     *
+     * @return loaded index.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public static final GZIIndex loadIndex(final Path indexPath) throws IOException {
+        if (indexPath == null) {
+            throw new IllegalArgumentException("null input path");
+        }
+        // allocate a buffer for re-use for read each byte
+        ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
+        buffer.order(ByteOrder.LITTLE_ENDIAN);
+
+        try (final ByteChannel channel = Files.newByteChannel(indexPath)) {
+            if (Long.BYTES != channel.read(buffer)) {
+                throw getCorruptedIndexException(indexPath, "less than " + Long.BYTES+ "bytes");
+            }
+            buffer.flip();
+
+            final int numberOfEntries;
+            try {
+                numberOfEntries = Math.toIntExact(buffer.getLong());
+            } catch (ArithmeticException e) {
+                buffer.flip();
+                throw getCorruptedIndexException(indexPath,
+                        String.format("HTSJDK cannot handle more than %d entries in .gzi index, but found %s",
+                                Integer.MAX_VALUE, buffer.getLong()));
+            }
+
+            // allocate array with the entries and add the first one
+            final List<IndexEntry> entries = new ArrayList<>(numberOfEntries + 1);
+            entries.add(FIRST_MAPPING);
+
+            // create a new buffer with the correct size and read into it
+            buffer = allocateBuffer(numberOfEntries, false);
+            channel.read(buffer);
+            buffer.flip();
+
+            for (int i = 1; i <= numberOfEntries; i++) {
+                final IndexEntry entry;
+                try {
+                    entry = new IndexEntry(buffer.getLong(), buffer.getLong());
+                } catch (IllegalArgumentException e) {
+                    throw new IOException(String.format("Corrupted index file: %s (%s)",
```

Comment:
I think it's always a good idea to pass the underlying exception in as the exception cause, even if your'e including the message in the top level exception.

Reply:
True, and actually this should use the static method for default formatting. Changed.